### PR TITLE
feat : implements acs 

### LINKED ISSUE #137: [acs] implements binary_map.go
implements binary_map and testcase

### PR DESCRIPTION
**Pull request by AgwaB**
## Related Issue

resolve: #137 #138 

## Description
I would like to discuss the following
- `acs.HandleInput`'s parameter should be []byte not `batch` becasue `HB` encrypt batch and send encrypted batch to `acs`

- How about remove `bba.Result` -> we have `binary channel`

- I wonder is it ok naming of interface `RBC` and `BBA`. i thought `rbc` or `bba` is just component, not actor.

- `tryAgreement`, `tryCompleteAgreement`, `processData`, `processAgreement` has no error return. because it is inner logic of `run` (channel and go routine). but i think it don't need to check error because already checked previous logic?

- many key of maps is `Address`. how about `Member`? `Member` has `Address` and `Member` can has more datas.

**After has some review, i will implements integration test on this request.**

## Checklist

Thank you for your contribution.
Before submitting this PR, please make sure:

- [x] Applying goimports
- [x] Test case
- [ ] End of Work

### CODE REVIEW

File: acs/acs.go [hunk 1]
```diff
@@ -35,40 +46,88 @@ type ACS struct {
...
+	binarySender cleisthenes.BinarySender,
+	batchSender cleisthenes.BatchSender,
+	broadCaster cleisthenes.Broadcaster,
+	coinGenerator cleisthenes.CoinGenerator) (*ACS, error) {
+
+	acs := &ACS{
+		n:                n,
+		f:                f,
+		owner:            owner,
+		rbcMap:           make(map[cleisthenes.Member]cleisthenes.RBC),
```
Comment (zeroFruit) [inline acs/acs.go]: abstract map plz

File: acs/acs.go [hunk 2]
```diff
@@ -116,20 +183,130 @@ func (acs *ACS) run() {
...
+			acs.tryCompleteAgreement()
 		}
 	}
 }
 
 // sendZeroToIdleBba send zero to bba instances which still do
 // not receive input value
-func (acs *ACS) sendZeroToIdleBba() {}
+func (acs *ACS) sendZeroToIdleBba() {
+	for addr, b := range acs.bbaMap {
```
Comment (zeroFruit) [inline acs/acs.go]: isn't `addr` actually `member`?

File: acs/acs.go [hunk 3]
```diff
@@ -116,20 +183,130 @@ func (acs *ACS) run() {
...
+		if state := acs.agreementStarted.item(addr); state.Undefined() {
+			b.HandleInput(&bba.BvalRequest{Value: cleisthenes.Zero})
+		}
+	}
+}
+
+func (acs *ACS) tryCompleteAgreement() {
+	if acs.dec.Value() && acs.countDoneAgreement() < acs.agreementThreshold() {
```
Comment (zeroFruit) [inline acs/acs.go]: isn't `!acs.dec.Value()`

File: acs/acs.go [hunk 4]
```diff
@@ -116,20 +183,130 @@ func (acs *ACS) run() {
...
+		return
+	}
+
+	agreementedNodeList := make([]cleisthenes.Member, 0)
```
Comment (zeroFruit) [inline acs/acs.go]: how about agreementedNodeList -> consensusedNodeList?

File: acs/acs.go [hunk 5]
```diff
@@ -116,20 +183,130 @@ func (acs *ACS) run() {
...
+}
 
-func (acs *ACS) tryCompleteAgreement(sender cleisthenes.Member, bin cleisthenes.Binary) {}
+func (acs *ACS) agreementSuccess(result map[cleisthenes.Member][]byte) {
+	acs.output = result
+	acs.batchSender.Send(cleisthenes.BatchMessage{})
+	acs.dec.Set(true)
+}
+
+func (acs *ACS) tryAgreement(sender cleisthenes.Member) {
```
Comment (zeroFruit) [inline acs/acs.go]: how about tryAgreement -> tryAgreementStart ?

File: acs/acs.go [hunk 6]
```diff
@@ -116,20 +183,130 @@ func (acs *ACS) run() {
...
+	bcResult := make(map[cleisthenes.Member][]byte)
+
+	for _, member := range agreementedNodeList {
+		if data := acs.broadcastResult.item(member); data != nil && len(data) != 0 {
+			bcResult[member] = data
+		}
+	}
+
+	if len(agreementedNodeList) == len(bcResult) && len(bcResult) != 0 {
+		acs.output = bcResult
```
Comment (zeroFruit) [inline acs/acs.go]: should send result to BatchReceiver

File: acs/acs.go (hunk 5 above)
Comment (AgwaB) [reply acs/acs.go]: good

File: acs/acs.go (hunk 2 above)
Comment (AgwaB) [reply acs/acs.go]: my mistake :(

**Review by zeroFruit — APPROVED** [review]
LGTM

**Pull request merged**